src/Microsoft.AspNetCore.Mvc.Razor/Internal/FileProviderRazorProjectItem.cs

```diff
@@ -27,11 +30,50 @@ public FileProviderRazorProjectItem(IFileInfo fileInfo, string basePath, string
 
         public override string PhysicalPath => FileInfo.PhysicalPath;
 
-        public override string RelativePhysicalPath { get; }
+        public override string RelativePhysicalPath
+        {
+            get
+            {
+                if (_relativePhysicalPath == null)
+                {
+                    if (Exists)
+                    {
+                        if (_root != null && !string.IsNullOrEmpty(PhysicalPath) && PhysicalPath.StartsWith(_root))
+                        {
+                            _relativePhysicalPath = PhysicalPath.Substring(_root.Length + 1); // Include leading separator
+                        }
+                        else
+                        {
+                            // Use FilePath if the file is not directly accessible
+                            _relativePhysicalPath = NormalizeAndEnsureValidPhysicalPath(FilePath);
+                        }
+                    }
+                }
+
+                return _relativePhysicalPath;
+            }
+        }
 
         public override Stream Read()
         {
             return FileInfo.CreateReadStream();
         }
+
+        private static string NormalizeAndEnsureValidPhysicalPath(string filePath)
+        {
+            if (string.IsNullOrEmpty(filePath))
+            {
+                return filePath;
+            }
+
+            filePath = filePath.Replace('/', Path.DirectorySeparatorChar);
+
+            if (filePath[0] == Path.DirectorySeparatorChar)
```

Comment:
Isn't this always true since you include the leading slash in your substring? Should we exclude that in the earlier substring? By the way, the other thing you could do is use my favorite type `InplaceStringBuilder` and avoid the extra sub-strings:


```C#
private string GenerateRelativePhysicalPath(string root, string path)
{
    Debug.Assert(path.StartsWith(root, StringComparison.OrdinalIgnoreCase));
    var builder = new InplaceStringBuilder(path.Length - root.Length);
    for (var i = root.Length; i < path.Length; i++)
    {
        var c = path[i];
        if (c == '/')
        {
            c = Path.DirectorySeparator;
        }
        
        builder.Append(c);
    }
    
    return builder.ToString();
}
```

Reply:
I was using this method to generate relative physical path from `FilePath` and not from `root` and `PhysicalPath`. Since we don't want to fallback to `FilePath` anymore, this method is not necessary.